pyk/src/pyk/kcfg/minimize.py

```diff
@@ -188,6 +200,106 @@ def _fold_lift(result: bool, finder_lifter: tuple[Callable, Callable]) -> bool:
+        """Merge targets of Split for cutting down the number of branches, using heuristics KCFGSemantics.is_mergeable.
+
+        Side Effect: The KCFG is rewritten by the following rewrite pattern,
+            - Match: A -|Split|-> A_i -|Edge|-> B_i
+            - Rewrite:
+                - if `B_x, B_y, ..., B_z are not mergeable` then unchanged
+                - if `B_x, B_y, ..., B_z are mergeable`, then
+                    - A -|Split|-> A_x or A_y or ... or A_z
+                    - A_x or A_y or ... or A_z -|Edge|-> B_x or B_y or ... or B_z
+                    - B_x or B_y or ... or B_z -|Split|-> B_x, B_y, ..., B_z
+
+        Specifically, when `B_merge = B_x or B_y or ... or B_z`
+        - `or`: fresh variables in places where the configurations differ
+        - `Edge` in A_merged -|Edge|-> B_merge: list of merged edges is from A_i -|Edge|-> B_i
+        - `Split` in B_merge -|Split|-> B_x, B_y, ..., B_z: subst for it is from A -|Split|-> A_1, A_2, ..., A_n
+        :param semantics: provides the is_mergeable heuristic
+        :return: whether any merge was performed
+        """
+        # ---- Match ----
+
+        # Step 1. Find all possible mergeable KCFG sub-graphs: A -|Split|> Ai -|Edge|> Bi
+        a2ai_list: list[KCFG.Split] = []  # A -|Split|> Ai
+        ai2bi_list: list[list[KCFG.Edge | KCFG.MergedEdge]] = []  # Ai -|Edge|> Bi
+        for split in self.kcfg.splits():
+            edges: list[KCFG.Edge | KCFG.MergedEdge] = [
+                single(self.kcfg.edges(source_id=ai)) for ai in split.target_ids if self.kcfg.edges(source_id=ai)
+            ]
+            edges = edges + [
+                single(self.kcfg.merged_edges(source_id=ai))
+                for ai in split.target_ids
+                if self.kcfg.merged_edges(source_id=ai)
+            ]
+            if len(edges) > 2:
+                a2ai_list.append(split)
+                ai2bi_list.append(edges)
+
+        # Step 3. Apply the heuristic & Obtain the merge-able KCFG sub-graphs
+        to_merge: dict[KCFG.Split, list[list[KCFG.Edge | KCFG.MergedEdge]]] = {}  # Split |-> Merge-able Edges
+        while a2ai_list:
+            a2ai = a2ai_list.pop()
+            ai2bi = ai2bi_list.pop()
+            while ai2bi:
+                mergeable_edges = [ai2bi.pop()]
+                idx = 0
+                while idx < len(ai2bi):
+                    if self.heuristics.is_mergeable(mergeable_edges[0].target.cterm, ai2bi[idx].target.cterm):
+                        for mergable_edge in mergeable_edges[1:]:
+                            if not self.heuristics.is_mergeable(mergable_edge.target.cterm, ai2bi[idx].target.cterm):
+                                raise ValueError(
+                                    'Mergeable edges are not partitioned, you should provide a better heuristic'
+                                )
+                        mergeable_edges.append(ai2bi.pop(idx))
+                    else:
+                        idx += 1
```

Comment:
I think we can factor out a function in `utils.py` (one directory up) called `partition_with` which does this:

```
def partition_with(relation: Callable[[T, T], bool], items: Iterable[T]) -> list[list[T]]:
    ...
```

That would make this routine a bit shorter and easier to read.

Reply:
Implementation:

https://github.com/runtimeverification/k/blob/28c3905319d56f11b417a4b76bc4406585856d02/pyk/src/pyk/utils.py#L322-L346

Test Cases:

https://github.com/runtimeverification/k/blob/28c3905319d56f11b417a4b76bc4406585856d02/pyk/src/tests/unit/test_utils.py#L63-L83

Application:

https://github.com/runtimeverification/k/blob/28c3905319d56f11b417a4b76bc4406585856d02/pyk/src/pyk/kcfg/minimize.py#L242-L248